src/aave-v2/lens/RatesLens.sol

```diff
@@ -359,7 +425,49 @@ abstract contract RatesLens is UsersLens {
         );
     }
 
-    /// INTERNAL ///
+    /// @notice Computes and returns the underlying pool rates for a specific market.
+    /// @param _underlying The underlying pool market address.
+    /// @param _supplied The amount hypothetically supplied.
+    /// @param _borrowed The amount hypothetically borrowed.
+    /// @param _repaid The amount hypothetically repaid.
+    /// @param _withdrawn The amount hypothetically withdrawn.
+    /// @return poolSupplyRate The market's pool supply rate per year (in ray).
+    /// @return poolBorrowRate The market's pool borrow rate per year (in ray).
+    function _getPoolRatesPerYear(
+        address _underlying,
+        uint256 _supplied,
+        uint256 _borrowed,
+        uint256 _repaid,
+        uint256 _withdrawn
+    ) internal view returns (uint256 poolSupplyRate, uint256 poolBorrowRate) {
+        DataTypes.ReserveData memory reserve = pool.getReserveData(_underlying);
+
+        PoolRatesVars memory vars;
+        (
+            vars.availableLiquidity,
+            vars.totalStableDebt,
+            vars.totalVariableDebt,
+            ,
+            ,
+            ,
+            vars.avgStableRate,
+            ,
+            ,
+
+        ) = dataProvider.getReserveData(_underlying);
+        (, , , , vars.reserveFactor) = reserve.configuration.getParamsMemory();
+
+        (poolSupplyRate, , poolBorrowRate) = IReserveInterestRateStrategy(
+            reserve.interestRateStrategyAddress
+        ).calculateInterestRates(
+            _underlying,
+            vars.availableLiquidity + _supplied + _repaid - _borrowed - _withdrawn,
+            vars.totalStableDebt,
+            vars.totalVariableDebt + _borrowed - _repaid,
+            vars.avgStableRate,
+            vars.reserveFactor
```

Comment:
:warning: The `_supplied` and `_borrowed` values take into account the supply and borrow of the user before the transaction. I think this shouldn't be the case as this supply (resp. borrow) has not moved into Aave because of the transaction. Only the added supply (resp. added borrow) should be taken into account, i.e. the increased supply amount line 86 in `Rates.sol` (res. increased borrow amount line 156)

Reply:
`getNextUserSupplyRatePerYear` calls:
```solidity
_getUserSupplyRatePerYear(
    _poolToken,
    balanceInP2P,
    balanceOnPool,
    _amount,
    repaidToPool
);
```

where `_amount` corresponds to what would be supplied to the pool (because if matching has been performed, `_amount` has been decreased accordingly)
and where `repaidToPool` corresponds to the amount matched peer-to-peer, if any

So I don't understand what you're trying to show me
However, I realized that `repaidToPool` & `withdrawnFromPool` were not accounting the matched delta

Reply:
Yes `_amount` and `repaidToPool` are correct I think. It's `balanceInP2P` and `balanceOnPool`, later named `_supplied` and `_borrowed`, that take into account the whole balance of the user. If I understand correctly, those 2 variables are supposed to simulate the fact that the pool is getting additional liquidity from the hypothetical transaction in `_getRatesPerYear` and in `_getPoolRatesPerYear`. The user can also have prior liquidity that is already accounted in the pool rate calculations, so we shouldn't pass the whole balance, but only the liquidity that is added on pool because of the hypothetical transaction

Reply:
Yes you are entirely right, I got the parameters mixed up